### PR TITLE
UCC/CL: cl basic starting implementation

### PR DESCRIPTION
**Pull request by vspetrov**
## What
 - Adds CL_BASIC component: build infra, lib_init/finalize, logging
 - Sets min required UCX version to 1.10
 - Minor change in ucc_framework: look for libucc_*.so objects instead of ucc_*.so (prefix lib appears due to libtool usage)

## Why ?
UCX 1.10 contains the UCS config table changes that used by UCC.

### CODE REVIEW

File: src/cl/basic/cl_basic_lib.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    ucc_status_t        status;
+    lib = (ucc_cl_basic_lib_t *)ucc_malloc(sizeof(*lib), "basic_lib");
+    ucc_cl_lib_init(&lib->super, &ucc_cl_basic.super, cl_config);
+    if (!lib) {
+        cl_error(&lib->super, "failed to allocate %zd bytes for lib object",
+                 sizeof(*lib));
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    lib->super.iface = &ucc_cl_basic.super;
```
Comment (alex--m) [inline src/cl/basic/cl_basic_lib.c:33]: This is clearly a case of subclassing - what do you say about using UCS_CLASS* macros?

**Review by alex--m — COMMENTED** [review]
Also, what's "ucc_tl_basic.h" for?

File: src/cl/basic/cl_basic_lib.c (hunk 1 above)
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c:33]: @alex--m yes, you are right, l'll check

File: src/cl/basic/cl_basic_lib.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+{
+    ucc_cl_basic_lib_t *lib = ucc_derived_of(cl_lib, ucc_cl_basic_lib_t);
+    cl_info(cl_lib, "finalizing lib object: %p", lib);
+    free(lib);
+    return UCC_OK;
+}
+
+ucc_cl_basic_iface_t ucc_cl_basic = {
+    .super.super.name = "basic",
+    .super.priority   = 10,
```
Comment (Sergei-Lebedev) [inline src/cl/basic/cl_basic_lib.c:52]: what is the difference between .super.super.priority and .super.priority?

File: src/cl/basic/cl_basic_lib.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    .super.cl_lib_config =
+        {
+            .name   = "CL_BASIC",
+            .prefix = "CL_BASIC_",
+            .table  = ucc_cl_basic_lib_config_table,
+            .size   = sizeof(ucc_cl_basic_lib_config_t),
+        },
+    .super.init     = ucc_basic_lib_init,
```
Comment (Sergei-Lebedev) [inline src/cl/basic/cl_basic_lib.c:60]: Maybe also set .super.super.type to UCC_CL_BASIC?

File: src/cl/basic/cl_basic_lib.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+#include "utils/ucc_malloc.h"
+
+static ucc_config_field_t ucc_cl_basic_lib_config_table[] = {
+    {"", "", NULL, ucc_offsetof(ucc_cl_basic_lib_config_t, super),
+     UCS_CONFIG_TYPE_TABLE(ucc_cl_lib_config_table)},
+
+    {NULL}
+};
+
+static ucc_status_t ucc_basic_lib_init(const ucc_lib_params_t *params,
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c:18]: i thought that we will not have tl "basic". TLs - transport layer - ucx, sharp, mcast, shmem, ...

File: src/cl/basic/cl_basic_lib.c (hunk 1 above)
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c]: added the subclassing for ucc_lib_* family of objects

File: src/cl/basic/cl_basic_lib.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                                       const ucc_lib_config_t *config,
+                                       const ucc_cl_lib_config_t *cl_config,
+                                       ucc_cl_lib_t **cl_lib)
+{
+    ucc_cl_basic_lib_t *lib;
+    ucc_status_t        status;
+    lib = (ucc_cl_basic_lib_t *)ucc_malloc(sizeof(*lib), "basic_lib");
+    ucc_cl_lib_init(&lib->super, &ucc_cl_basic.super, cl_config);
+    if (!lib) {
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c]: switched to UCC_CLASS* (it checks for oom)

File: src/cl/basic/cl_basic_lib.c (hunk 2 above)
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c:52]: iirc base component iface does not have a priority field, should not be super.super.priority.

File: src/cl/basic/cl_basic_lib.c (hunk 3 above)
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c:60]: yes

File: src/cl/basic/cl_basic_lib.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (manjugv) [reply src/cl/basic/cl_basic_lib.c:34]: Though we will not have tl basic component, I think it make sense to use ucc_cl_basic namespace.

**Review by manjugv — APPROVED** [review]

**Comment by manjugv** [comment]
@vspetrov It seems like CI is complaining about commit title. It seems like it is working. :) UCC/CL -> CL should work.

**Review by alex--m — APPROVED** [review]

File: src/cl/basic/cl_basic_lib.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c]: done

**Pull request merged**